shell/platform/common/cpp/text_input_model_unittests.cc

```diff
@@ -0,0 +1,364 @@
+
+#include "gtest/gtest.h"
+
+namespace flutter {
+
+TEST(TextInputModel, InputTypeAndAction) {
+  auto model = std::make_unique<TextInputModel>("INPUT_TYPE", "INPUT_ACTION");
+  EXPECT_STREQ(model->input_type().c_str(), "INPUT_TYPE");
+  EXPECT_STREQ(model->input_action().c_str(), "INPUT_ACTION");
+}
+
+TEST(TextInputModel, SetEditingStateStart) {
+  auto model = std::make_unique<TextInputModel>("", "");
+  EXPECT_TRUE(model->SetEditingState(0, 0, "ABCDE"));
+  EXPECT_EQ(model->selection_base(), 0);
+  EXPECT_EQ(model->selection_extent(), 0);
+  EXPECT_STREQ(model->GetText().c_str(), "ABCDE");
+}
+
+TEST(TextInputModel, SetEditingMiddle) {
+  auto model = std::make_unique<TextInputModel>("", "");
+  EXPECT_TRUE(model->SetEditingState(2, 2, "ABCDE"));
+  EXPECT_EQ(model->selection_base(), 2);
+  EXPECT_EQ(model->selection_extent(), 2);
+  EXPECT_STREQ(model->GetText().c_str(), "ABCDE");
+}
+
+TEST(TextInputModel, SetEditingStateEnd) {
+  auto model = std::make_unique<TextInputModel>("", "");
+  EXPECT_TRUE(model->SetEditingState(5, 5, "ABCDE"));
+  EXPECT_EQ(model->selection_base(), 5);
+  EXPECT_EQ(model->selection_extent(), 5);
+  EXPECT_STREQ(model->GetText().c_str(), "ABCDE");
+}
+
+TEST(TextInputModel, SetEditingStateSelection) {
+  auto model = std::make_unique<TextInputModel>("", "");
+  EXPECT_TRUE(model->SetEditingState(1, 4, "ABCDE"));
+  EXPECT_EQ(model->selection_base(), 1);
+  EXPECT_EQ(model->selection_extent(), 4);
+  EXPECT_STREQ(model->GetText().c_str(), "ABCDE");
+}
+
+TEST(TextInputModel, SetEditingStateReverseSelection) {
+  auto model = std::make_unique<TextInputModel>("", "");
+  EXPECT_FALSE(model->SetEditingState(3, 1, "ABCDE"));
+}
+
+TEST(TextInputModel, SetEditingStateOutsideString) {
+  auto model = std::make_unique<TextInputModel>("", "");
+  EXPECT_FALSE(model->SetEditingState(4, 6, "ABCDE"));
+  EXPECT_FALSE(model->SetEditingState(5, 6, "ABCDE"));
+  EXPECT_FALSE(model->SetEditingState(6, 6, "ABCDE"));
+}
+
+TEST(TextInputModel, SetEditingStateWideCharacters) {
+  auto model = std::make_unique<TextInputModel>("", "");
+  EXPECT_TRUE(model->SetEditingState(0, 0, "😄🙃🤪🧐"));
+  EXPECT_EQ(model->selection_base(), 0);
+  EXPECT_EQ(model->selection_extent(), 0);
+  EXPECT_STREQ(model->GetText().c_str(), "😄🙃🤪🧐");
+}
+
+TEST(TextInputModel, AddCodePoint) {
+  auto model = std::make_unique<TextInputModel>("", "");
+  model->AddCodePoint('A');
+  model->AddCodePoint('B');
+  model->AddCodePoint('C');
+  model->AddCodePoint('D');
+  model->AddCodePoint('E');
+  EXPECT_EQ(model->selection_base(), 5);
+  EXPECT_EQ(model->selection_extent(), 5);
+  EXPECT_STREQ(model->GetText().c_str(), "ABCDE");
+}
+
+TEST(TextInputModel, AddCodePointSelection) {
+  auto model = std::make_unique<TextInputModel>("", "");
+  EXPECT_TRUE(model->SetEditingState(1, 4, "ABCDE"));
+  model->AddCodePoint('x');
+  EXPECT_EQ(model->selection_base(), 2);
+  EXPECT_EQ(model->selection_extent(), 2);
+  EXPECT_STREQ(model->GetText().c_str(), "AxE");
+}
```

Comment:
Same here, making sure the selection point updates by the correct amount for a multi-unit code point.